### PR TITLE
Raise exception on out-of-bounds tensor indexing

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108910
* #108690

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1209,8 +1209,19 @@ def scale_index_with_offset(
...
+        generate_assert = (
+            (check or config.debug_index_asserts)
+            and config.cpp.assert_indirect_indexing
+        )
+
+        if generate_assert:
+            expr = ops.and_(ops.ge(index_var, '0'),
+                            ops.le(index_var, self.rename_indexing(size)))
+            cond = self.cse.generate(self.compute, expr)
+            line = f'assert({cond} && "index out of bounds");'
```
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:1222]: Rather than doing a naked assert, please use `TORCH_CHECK`.

File: torch/_inductor/config.py [hunk 2]
```diff
@@ -306,6 +306,9 @@ class cpp:
     # how many nodes to allow into a single horizontal fusion
     max_horizontal_fusion_size = 16
 
+    # assert that indirect indexing does not read / write out of bounds
+    assert_indirect_indexing = True
+
```
Comment (lezcano) [inline torch/_inductor/config.py:311]: We should probably have a shared config for CPU and CUDA

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -1209,8 +1209,19 @@ def scale_index_with_offset(
...
+            self.compute.writeline(line)
```
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:1223]: Could we also share the logic where, if we find a tighter bound on a variable later on, we elide the check?

**Review by lezcano — COMMENTED** [review]
You'll need to add in this PR the line testing that the generated code has indeed the `TORCH_CHECK` on CPU

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:1222]: Also, do you need to handle `indirect_max_sizes` the similar way as triton here?

Comment (lezcano) [reply torch/_inductor/codegen/cpp.py:1222]: Yep, that's https://github.com/pytorch/pytorch/pull/108910#discussion_r1320610435. Note that that's an optional optimisation, but it'd be good to have now that we're at it.

**Comment by guilhermeleobas** [comment]
Closing this one in favor of https://github.com/pytorch/pytorch/pull/108690

**Pull request closed without merging**